### ISSUE TITLE
Enhancement: [strict-boolean-expressions] allowOptionalChaining

### ISSUE THREAD

**Issue by rivatove** [typescript-eslint/typescript-eslint#8694]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md

### Description

From my point of view, when I optionally chain a value, I'm explicitly stating that I understand it can be null/undefined apart from the base type.

I suggest an extra option `allowOptionalChaining` that would make the rule not throw an error in such cases.

### Fail

```TypeScript
// The code from the Pass section should not report an error, which is the current behavior.
```

### Pass

```TypeScript
// Example #1
// date1 & date2 are Dayjs objects. date1 is nullable.
date1?.isBefore(date2)

// Example #2
if (options?.explicitDiscard) {
   do stuff...
}
```

**Comment by bradzacher** [comment]
I'm not sure I understand or agree with the statement

> when I optionally chain a value, I'm explicitly stating that I understand it can be null/undefined apart from the base type.

How is a `string | undefined` obtained from a variable any different to one obtained from an optional chain?

They're the same type and using a loose falsey check has exactly the same result. Why should one be treated differently to the other?

**Comment by rivatove** [comment]
> How is a `string | undefined` obtained from a variable any different to one obtained from an optional chain?

When I scan such code, it seems transparent to me that I'm evaluating a boolean with a default falsy value. Because the optional chaining is so explicit visually.

I'm aware I may be an outlier and this is not the way people usually look at it. But I think it's worth asking the community.

**Comment by bradzacher** [comment]
Put it another way
```ts
if (x?.y) {}

const r = x?.y;
if (r) {}
```

Why should these two code fragments be handled differently by the rule?

Given this is the first request for such a feature in the many years the rule has existed - it's not something most people consider like you do - because people consider types here, not syntax based exceptions.

Without a compelling argument to make this change, I'm going to close this.
